### ISSUE TITLE
x/tools/gopls: regtest flakes due to hanging go commands

### ISSUE THREAD

**Issue by findleyr** [golang/go#54461]
This recent flake on the build dashboard looks like a clear deadlock during shutdown:

> greplogs --dashboard -md -l -e 'runCmdContext' --since=2022-07-14

<details>
[2022-08-15T17:42:12-987de34-1f833e4/darwin-amd64-12_0](https://build.golang.org/log/00046e0b005c7660d676a3a415561950048f756a)
[2022-08-14T00:06:23-35f806b-59865f1/netbsd-amd64-9_0](https://build.golang.org/log/cac6093c9a1e29b020d1a0076f0b67e262a134c5)
[2022-08-12T20:40:05-bebd890-2f6783c/netbsd-386-9_0](https://build.golang.org/log/3680daecab8adb2905484af9298a4239b3812f6a)
[2022-08-12T18:15:28-bebd890-b6f87b0/netbsd-amd64-9_0](https://build.golang.org/log/c2900e262adb237ea981093befa87ec35faa804b)
[2022-08-12T12:39:26-88d981e-f67c766/netbsd-amd64-9_0](https://build.golang.org/log/d22326763f8cbedb39bb3c17f3e66e57efa453df)
[2022-08-12T00:04:29-c4ec74a-a5cd894/netbsd-386-9_0](https://build.golang.org/log/eec2b475f968a73f4f17745f72013ca6b3dc2a26)
[2022-08-11T19:05:54-c4ec74a-62654df/netbsd-amd64-9_0](https://build.golang.org/log/070b4fee818231324badb590600980000d349fc2)
[2022-08-11T17:53:50-37a81b6-a526ec1/netbsd-amd64-9_0](https://build.golang.org/log/373391707f93fff54945c55648170d9dd390958c)
[2022-08-11T16:19:14-37a81b6-2340d37/netbsd-amd64-9_0](https://build.golang.org/log/1c1abf0fdf7307ffbe030628c0c37a2af50e044d)
[2022-08-11T12:53:59-b2156b5-3c200d6/netbsd-386-9_0](https://build.golang.org/log/dd3e35eeace9e0ab8bad2da68b3822e492084f55)
[2022-08-10T22:22:48-b2156b5-6b80b62/netbsd-amd64-9_0](https://build.golang.org/log/0e0cf575d88b65e91a55b8f070bf9b7ef3136e6e)
[2022-08-10T17:41:25-0ad49fd-f19f6c7/netbsd-amd64-9_0](https://build.golang.org/log/f1c0a3ff9bc966ba2ea145a3fc5b207e7da04422)
[2022-08-10T15:08:24-3950865-c81dfdd/netbsd-amd64-9_0](https://build.golang.org/log/49e68c6f9a3e02fea5a3560b3599fec748fe37a5)
[2022-08-10T02:14:09-6fa767d-5531838/plan9-386-0intro](https://build.golang.org/log/dd0540ff0994f1f4dabc184f1a3139bc92e459b0)
[2022-08-09T14:33:24-92d58ea-0981d9f/openbsd-386-70](https://build.golang.org/log/2257a9730544b80a3bbd32cd77d842bae30df216)
[2022-08-09T14:12:01-92d58ea-662a729/netbsd-amd64-9_0](https://build.golang.org/log/262bb87feffe594abd0bca78d7ff0762bce6ce6d)
[2022-08-09T13:39:27-92d58ea-9e8020b/netbsd-386-9_0](https://build.golang.org/log/5038ff991c57474969f20b56f7e0147b970f9d6f)
[2022-08-09T11:28:56-92d58ea-0f8dffd/netbsd-amd64-9_0](https://build.golang.org/log/6f487445444131ce88dc81395c61491a75b717b6)
[2022-08-08T18:10:56-fff6d6d-4bcc138/netbsd-amd64-9_0](https://build.golang.org/log/d7b15610ae881f54e8580b1df112b35fae495416)
[2022-08-08T15:33:45-06d96ee-0581d69/netbsd-amd64-9_0](https://build.golang.org/log/0f6a50f7330a08b10fbbb2b30592206aa78088d8)
[2022-08-08T15:07:46-06d96ee-cd54ef1/netbsd-amd64-9_0](https://build.golang.org/log/4bf9c1e647958f6f7e91f6a4af54f5c0d47a5ff9)
[2022-08-08T14:12:21-06d96ee-e761556/netbsd-amd64-9_0](https://build.golang.org/log/139292d4921bc45d1d4f7eaad29b8ab74d492337)
[2022-08-08T06:16:59-06d96ee-0f6ee42/darwin-amd64-11_0](https://build.golang.org/log/dcb2262105851e705bd8a54363d0f7eb12ced52d)
[2022-08-08T06:16:59-06d96ee-0f6ee42/netbsd-386-9_0](https://build.golang.org/log/e338e8cc5f896780178f4653ca14cb3c5420eee3)
[2022-08-06T15:20:00-06d96ee-0c4db1e/plan9-386-0intro](https://build.golang.org/log/5cca74b05c0a96fe84d714c05095b688cf142998)
[2022-08-05T19:51:08-06d96ee-4fb7e22/plan9-386-0intro](https://build.golang.org/log/6b5544e536b89cd2b350dc28831b3cf2c1b930d1)
[2022-08-04T20:05:03-81c7dc4-39728f4/netbsd-386-9_0](https://build.golang.org/log/a0e0e2e24dbcc85bc1112cde350e2aab2f68507e)
[2022-08-04T20:05:03-81c7dc4-39728f4/netbsd-amd64-9_0](https://build.golang.org/log/1cf98e0db8d30c95362f07d44c47c5363b8c3955)
[2022-08-04T20:04:16-3519aa2-39728f4/netbsd-386-9_0](https://build.golang.org/log/9395fd87814bc4d6344a3bbf72f427637a40ff9d)
[2022-08-04T19:57:25-763f65c-39728f4/netbsd-386-9_0](https://build.golang.org/log/e6a1d9cdf59a37ef2641727a09850df931bf6c76)
[2022-08-04T18:51:46-99fd76f-39728f4/openbsd-386-70](https://build.golang.org/log/d3aa446a547efa10e17e2f9096f34a1e7de84a34)
[2022-08-04T17:05:18-3e0a503-fb1bfd4/netbsd-amd64-9_0](https://build.golang.org/log/3eab42b91ac882d28586f90d833b6a7adadfaa8e)
[2022-08-04T15:50:11-3e0a503-fcdd099/netbsd-386-9_0](https://build.golang.org/log/f20346aaff38f3501b9586c4bd5bce30beeb544c)
[2022-08-04T15:50:11-3e0a503-44ff9bf/netbsd-amd64-9_0](https://build.golang.org/log/cc32154c3c01fea6178c088a401ce789b8cdfea2)
[2022-08-04T15:31:49-87f47bb-44ff9bf/plan9-386-0intro](https://build.golang.org/log/1a849764f69705f80e2b9215a4af08a35ca6b599)
[2022-08-04T14:58:59-87f47bb-4345620/netbsd-386-9_0](https://build.golang.org/log/1979e42d0a685eaba423a2b72abb5f01467f9d29)
[2022-08-04T10:32:51-3e0a503-a10afb1/linux-386-buster](https://build.golang.org/log/0575b7d9be6333dd398544d939a632c7bca98820)
[2022-08-03T21:02:27-8b9a1fb-f28fa95/plan9-386-0intro](https://build.golang.org/log/3bdc65bf0d8317c8060a9213e294bd69d8c9a2c6)
[2022-08-03T21:02:27-8b9a1fb-4345620/netbsd-386-9_0](https://build.golang.org/log/274203fd71753a2475dc09b150bf2223ccafc467)
[2022-08-03T18:07:40-d08f5dc-fcdd099/netbsd-386-9_0](https://build.golang.org/log/0356b836f12656f38f905a8b00195c0dc91f9f9c)
[2022-08-03T13:50:38-ddb90ec-c6a2dad/dragonfly-amd64-622](https://build.golang.org/log/2afbb152ac08399d2e748415c1e8237aba177f93)
[2022-08-03T13:50:38-ddb90ec-c6a2dad/plan9-386-0intro](https://build.golang.org/log/17a5aa41dccbd8d4d12f725824e944297efc759d)
[2022-08-03T12:09:24-ddb90ec-29b9a32/plan9-386-0intro](https://build.golang.org/log/b476964cf820ebb56acaf4873de238e9944c2fdc)
[2022-08-02T18:52:36-0d04f65-29b9a32/plan9-386-0intro](https://build.golang.org/log/8d8b0cf1e7abad05f3b5efc3d81f8b943459e208)
[2022-08-02T18:19:01-d025cce-be59153/netbsd-amd64-9_0](https://build.golang.org/log/d373872d8833fcd7943ed0c44a350b1ed0964171)
[2022-08-02T18:16:22-10cb435-d723df7/netbsd-amd64-9_0](https://build.golang.org/log/52dcc80905a7e4fb69b303d2e10b0853d7a4d249)
[2022-08-02T18:07:14-4d0b383-d723df7/netbsd-386-9_0](https://build.golang.org/log/32e87f0ad4c401d7b6898923db1ccf0e18e96a55)
[2022-08-02T18:07:14-4d0b383-d723df7/netbsd-amd64-9_0](https://build.golang.org/log/c45079cd676d8c78a4fb2e516c7f8c61e99cf932)
[2022-08-02T17:23:42-4d0b383-1b7e71e/darwin-amd64-nocgo](https://build.golang.org/log/f1e5f794c6590cf6ef6af5b32593bdb2dc4e691a)
[2022-08-02T16:05:48-4d0b383-f2a9f3e/netbsd-amd64-9_0](https://build.golang.org/log/615a1fe11bff97f99eea28f57de7379a57de05ff)
[2022-07-29T20:19:23-9580c84-9240558/windows-arm64-11](https://build.golang.org/log/902f203f4ca2a092814c8f9feae9f34403c03fc0)
[2022-07-28T20:06:00-8ea5687-d9242f7/darwin-amd64-nocgo](https://build.golang.org/log/12a2b665a640b80267a18e005b8f48cef599747b)
[2022-07-27T15:04:58-39a4e36-4248146/freebsd-386-13_0](https://build.golang.org/log/2285dd340580856181ca3d5dc0c60759238a178d)
[2022-07-26T18:43:08-6c8a6c4-d9242f7/aix-ppc64](https://build.golang.org/log/b396e3df87257d69b4dfdaf4a5f927b2109cb3f4)
[2022-07-25T20:44:49-2a6393f-24dc27a/darwin-amd64-10_14](https://build.golang.org/log/99081c715af7f107984bbf25ee529a1bb386afe1)
[2022-07-25T18:11:01-4375b29-795a88d/plan9-386-0intro](https://build.golang.org/log/cb5fe95e5d275800ef3139747fb602ad99ea159e)
[2022-07-25T14:16:17-178fdf9-64f2829/plan9-386-0intro](https://build.golang.org/log/d5eae87730ea926b99e5c20558b3f7d53adcaa7e)
[2022-07-22T20:12:19-7b605f4-c5da4fb/plan9-386-0intro](https://build.golang.org/log/745ec57924a1086a4044a42e8daaf9cdad4b755a)
[2022-07-21T20:11:06-ec1f924-c4a6d30/plan9-386-0intro](https://build.golang.org/log/4a72300281f575745ce47d3a3e719e2ad624395f)
[2022-07-15T15:11:26-22d1494-2aa473c/windows-386-2008-newcc](https://build.golang.org/log/76a64e26cc4949d727458d3d75bf0a7d957761ba)
[2022-07-15T14:27:36-1a4e02f-4651ebf/windows-arm64-10](https://build.golang.org/log/31cefa7b619aab91a17a2e86457aea315f08992f)
[2022-07-15T14:20:24-db8f89b-4651ebf/windows-arm64-10](https://build.golang.org/log/32b673e2683994bcdf4b3ca0031b2ba66d6fa5f2)
[2022-07-14T21:03:14-db8f89b-783ff7d/windows-arm64-11](https://build.golang.org/log/da6886000059b7db0800204df81e5ae4cd151a5b)
[2022-07-14T21:01:58-db8f89b-aa80228/darwin-arm64-11](https://build.golang.org/log/cbff206e69443d9ec55c0fb38c5b0f3205fe7ba9)
[2022-07-14T19:05:09-db8f89b-a906d3d/windows-arm64-10](https://build.golang.org/log/283ac24f2b8085b434b86ff5fa37d5762efcaa40)
[2022-07-14T15:54:36-db8f89b-266c70c/windows-arm64-10](https://build.golang.org/log/9b4cda8292abc11d4807641e8af88e1cf57b2fd3)
[2022-07-14T01:47:39-db8f89b-558785a/windows-arm64-11](https://build.golang.org/log/d9dc888e4b2ad7d0a1b9804a1dd87cc9e4329946)
</details>

We recently started waiting for all go command invocations when shutting down gopls regtests.  It appears that sometimes we kill the go command and still don't get a result from `cmd.Wait()`. For example, here:
https://build.golang.org/log/00046e0b005c7660d676a3a415561950048f756a

In that failure, the test runner looks otherwise healthy (other tests ran fast), and yet the goroutine stack clearly shows a go command hanging here:
https://cs.opensource.google/go/x/tools/+/master:internal/gocommand/invoke.go;l=260;drc=f38573358cbedf46d64c5759ef41b72afcf0c5c0

@bcmills do you happen to have any idea of what might cause this?

**Comment by findleyr** [comment]
Two observations:

1. We're not checking the error from `Kill`. If killing failed, I don't think we can expect Wait to return.
2. Discussion at https://groups.google.com/g/golang-nuts/c/pUaEjqd5wo8

**Comment by adonovan** [comment]
The only reason the kill system call can fail (at least in this situation) is when the child process has already exited, so failure of kill is unlikely to be the culprit. More likely kill terminated the go process itself, but not the tree of processes rooted at it. If one of them (a test?) retains an open file descriptor to the stdout pipe created by os/exec then the cmd.Run operation will hang indefinitely.

**Comment by bcmills** [comment]
Please do check the error from `Kill` — that would at least give a useful data point for diagnostics. (This may be closely related to #53400.)

**Comment by findleyr** [comment]
Interesting, I was debugging this in https://go.dev/cl/424075. On windows, our call to Process.Kill() is failing with "invalid argument":
https://storage.googleapis.com/go-build-log/e2f60200/windows-386-2008_71ad7007.log

A bit of googling suggests that this is because we can't kill subprocesses on windows.

@bcmills any advice for how to properly kill the go command on windows?

**Comment by findleyr** [comment]
After reading the source a bit more: this is EINVAL, which appears to mean that the Process.wait() has exited and the handle released, so this is a race, although it is surprising that we hit it so reliably.

**Comment by bcmills** [comment]
> any advice for how to properly kill the go command on windows?

Can't be done without creating a whole extra process group, unfortunately. (Probably we should add a side-channel — perhaps an open file descriptor or a pidfile? — to request clean shutdown on Windows.)

**Comment by findleyr** [comment]
A hit!
https://build.golang.org/log/4eea0e5bfe425fcd097a49fcb92fd015fa88f383

**Comment by adonovan** [comment]
Nice. Well, that test process seems very much alive, falsifying my hypothesis.

**Comment by findleyr** [comment]
```
13032 12910 (compile)
 9802 13032 go list -modfile=/tmp/workdir/tmp/go.8f5d4cc245da9790bda529e64ae3e7
 ```
 
Looks like the hanging go command is in the middle of a compile. Wish we had the full subprocesss command line -- I'll look into that.

Not sure how to interpret the fstat output.

**Comment by adonovan** [comment]
That's a dead cmd/compile process: there's no command because argv has been destroyed along with the rest of the address space. Perhaps the go list parent simply hasn't called waitpid yet, so the process table entry has to be retained. I suspect the problem is in go list.

**Comment by findleyr** [comment]
Aha, thanks (excuse my ps noobness).

Note that we instrumented this panic in two places: once before `Kill()` and once after. This is _before_, so I don't think it falsifies your hypothesis, if I'm understanding correctly.

**Comment by findleyr** [comment]
`greplogs --dashboard -md -l -e 'HANGING GO' --since=2022-09-15`

[2022-09-17T02:56:51-4d18923-cc1b20e/netbsd-amd64-9_0](https://build.golang.org/log/333cfe1792860f9ae99158d4b91be4de74ee565b)
[2022-09-16T17:59:31-a61f20e-a84f46a/netbsd-amd64-9_0](https://build.golang.org/log/caf074922535c82086b3055373afed4eb7c852d6)
[2022-09-16T16:33:11-a61f20e-8451529/netbsd-386-9_0](https://build.golang.org/log/df54908eba0ab6e17498c82572922785dd399591)
[2022-09-16T14:49:13-a61f20e-b35b4a3/netbsd-amd64-9_0](https://build.golang.org/log/4eea0e5bfe425fcd097a49fcb92fd015fa88f383)

Still only netbsd. Posting the `greplogs` query here for future reference.

**Comment by bcmills** [comment]
Ooh, nice! https://go.dev/issue/55323#issuecomment-1254107802 has a `darwin/arm64` hang. 😃

**Comment by bcmills** [comment]
As expected, `cmd/go` itself has already exited at the point of the hang:
```
2022/09/20 08:42:37 error killing the Go command: os: process already finished
```

But there are an awful lot of processes in that dump, and nothing that obviously stands out as a subprocess of `cmd/go` that would have access to `stdout` or `stderr`. 😕

**Comment by bcmills** [comment]
Here's an off-the-wall theory: maybe the file descriptors actually *are* closed, and something in the parent Go process (i.e. `gopls`) isn't waking up from the end-of-stream event..?

#53434 could be related.

**Comment by bcmills** [comment]
To check that hypothesis, two suggestions:
1. `gopls` should log exactly which PID was the stuck `go` command, just in case it left a remnant in the `ps` or` lsof` output that we can track down.
2. `gopls` should invoke `debug.SetTraceback("system")` before the terminal `panic` [here](https://cs.opensource.google/go/x/tools/+/master:internal/gocommand/invoke.go;l=333;drc=a61f20e1aa172e30522c65973ead5509cecf01c2).

**Comment by adonovan** [comment]
A simpler explanation is that gopls sent SIGINT, waited for 1s, then sent SIGKILL, and the child (go) exited during the 1s grace period, causing kill(2) to fail with ESRCH (-> ErrProcessDone = "process already finished"). Seems like that's to be expected if a busy machine causes a delay in responding to SIGINT. Perhaps the debugging logic should treat ErrProcessDone as success, not a reason to debug?

**Comment by findleyr** [comment]
@adonovan but in this case the panic indicates that we waited up to 10 seconds *after* that kill. The error from Kill is merely logged.

Per @bcmills comment on the CL, we should expect that `cmd.Wait` returns quickly once the process is done.

**Comment by adonovan** [comment]
> the panic indicates that we waited up to 10 seconds after that kill. The error from Kill is merely logged.

Good point.

So the sequence of events that we know must have happened from the log is:
- ctx cancelled
- SIGINT
- wait 1s
- SIGKILL => ErrProcessDone
- wait 10s
- dump shows a go process still alive

ErrProcessDone means that either waitpid returned ESRCH or that this was the second call to Process.signal after a first one set the 'done' event at the end of a successful call to Process.wait. In either case, the process must already have terminated.

Is it possible that the dump shows a different go process? Seems improbable, but I feel like we've eliminated the impossible. We should log the pid of cmd.Process, like @bcmills suggested.

**Comment by adonovan** [comment]
> Is it possible that the dump shows a different go process? Seems improbable, but I feel like we've eliminated the impossible. We should log the pid of cmd.Process, like @bcmills suggested.

The pid in the last log above clearly shows that that the go list process is the child of the test, not some other process in the container. Huh...

**Comment by adonovan** [comment]
We still haven't seen a single log with both (a) the "error killing the Go command" message and (b) a pid showing that 'go' is the child of the test. So it's possible that the log with just (a) was merely a very slow (netbsd) machine and the log with just (b) was showing a different go process in the ps output. If both were to occur in the same log it would indicate a bug in the system call.

**Comment by bcmills** [comment]
> merely a very slow (netbsd) machine

The NetBSD machines aren't particularly slow; they're buggy (#50138 up until yesterday, and #56180 after that).

**Comment by adonovan** [comment]
From the last log, we're seeing that `go list` (pid 11323) is slow to complete, more than 1 minute, on the netbsd builder. We don't have the complete go list command, but we can tell from the output of ps that it is definitely running the compiler, which is surprising, as that should only be needed in -export mode, which gopls has no need for. Notice the three "(compile)" child processes in ps, plus all gopls' in/out/err pipes in the lsof dump. The compile processes have all terminated (ps cannot display their argv arrays because the address spaces have been destroyed), yet the go list process has not closed its pipes.

```
ps axo ppid,pid,command:
-------------------------
 PPID   PID COMMAND
...
  314  6958 /tmp/workdir/go/bin/go test -short ./... 
 6958 10212 /tmp/workdir/tmp/go-build957980754/b501/completion.test -test.testl
11323 10917 (compile)
10212 11323 go list -modfile=/tmp/workdir/tmp/go.e8b79c677cbf58581b70654d24944f [truncated]
11323 11598 (compile)
11323 11901 (compile)
...
```


fstat:
-----
USER     CMD          PID   FD MOUNT       INUM MODE         SZ|DV R/W
...
root     go         11323   wd  /         1622282 drwxr-xr-x     512 r 
root     go         11323    0  /         1713612 crw-rw-rw-    null r 
root     go         11323    1* pipe 0xffff80320e6911a0 -> 0xffff803233ae22c0 w
root     go         11323    2* pipe 0xffff8031e6711d98 -> 0xffff803233ae2ea0 w
root     go         11323    3  /         1400146 -rw-r--r--   86547 r 
root     go         11323    4  /         1442046 -rw-r--r--    1115 r 
root     go         11323    5  /         1448217 -rw-r--r--     186 r 
root     go         11323    6  /         1442053 -rw-r--r--    3407 r 
root     go         11323    7  /         1442035 -rw-r--r--     195 r 
root     go         11323    8  /         1442034 -rw-r--r--    5573 r 
root     go         11323    9  /         1442142 -rw-r--r--    3069 r 
root     go         11323   10  /         1442405 -rw-r--r--     133 r 
root     go         11323   11  /         1442062 -rw-r--r--    2864 r 
root     go         11323   12  /         1442418 -rw-r--r--    1972 r 
root     go         11323   13  /         1442427 -rw-r--r--    3924 r 
root     go         11323   14  /         1348570 -rw-r--r--     393 r 
root     go         11323   15  /         1442421 -rw-r--r--    1036 r 
root     go         11323   16* pipe 0xffff8031a85218d0 <- 0x0 rn
root     go         11323   17* kqueue pending 0
root     go         11323   18* pipe 0xffff803253411638 <- 0xffff80325f6a22a8 rn
root     go         11323   19* pipe 0xffff80325f6a22a8 -> 0xffff803253411638 wn
root     go         11323   20* pipe 0xffff8032539119d0 <- 0x0 rn
root     go         11323   21* pipe 0xffff8030f7b30c70 <- 0x0 rn
root     go         11323   22* pipe 0xffff803283de7e80 <- 0x0 rn
root     go         11323   23* pipe 0xffff8031d28892d8 <- 0x0 rn
root     go         11323   24  /         1399130 -rw-r--r--     329 r 
root     go         11323   27* pipe 0xffff8030f7b30b40 <- 0x0 rn
root     go         11323   28* pipe 0xffff80319abc1c68 <- 0x0 rn
root     go         11323   29* pipe 0xffff8030db831e80 <- 0x0 rn
root     go         11323   34* pipe 0xffff803233637c48 <- 0x0 rn
root     go         11323   36  /         1713612 crw-rw-rw-    null r 
root     go         11323   37* pipe 0xffff8030f7b30090 <- 0xffff80304c451af0 rn
root     go         11323   38* pipe 0xffff80304c451af0 -> 0xffff8030f7b30090 wn
root     go         11323   39* pipe 0xffff8031a8521670 <- 0xffff80304c451170 rn
root     go         11323   40* pipe 0xffff80304c451170 -> 0xffff8031a8521670 wn
...
```

**Comment by adonovan** [comment]
Yes, that does look like the explanation. (Nice work tracking that one down, BTW.)

I'll limit my attention to non-netbsd platforms going forward.

**Comment by findleyr** [comment]
Note: some recent flakes have been in darwin (rather than netbsd), and may contain interesting results. (Haven't dug in yet)

**Comment by bcmills** [comment]
The `netbsd-386-9_3` hangs look like a platform bug to me (filed as #57999).

**Comment by bcmills** [comment]
Looking at the recent `darwin` failures, they all appear to have occurred with the `go` command already exited.

I wonder if the check is too sensitive at this point? The examples in the original post were hanging for multiple minutes, not just a few seconds after the test. 🤔

**Comment by adonovan** [comment]
There have been a lot of reports of this lately, and we shouldn't even be running gopls tests on a 386 (or any 32-bit machine).

**Comment by bcmills** [comment]
I don't think we've seen any failures on a `386` platform lately?

They're mostly on the `darwin/amd64` platform, and with the failure mode of `error killing the Go command: os: process already finished`, which suggests to me that the 1s timeout after `cmd.Process.Kill()` is probably just too short:
https://cs.opensource.google/go/x/tools/+/master:internal/gocommand/invoke.go;l=277;drc=5214f412aecffbca887fc0c26f8d59e3a2342d44

**Comment by adonovan** [comment]
> I don't think we've seen any failures on a 386 platform lately?

Sorry, ignore me. Scanning too many watchflakes posts in haste, I misread the exclusion of 386 as a positive report.

> 1s ... is probably just too short

Yeah, that seems plausible. Shall I make it 2? ;-)

**Comment by bcmills** [comment]
Thinking about this some more: if the process is already finished but `Wait` is taking longer than a second, it's probably spending that time ingesting a lot of (presumably buffered) output from the command's `stdout` and/or `stderr`.

Perhaps we should be redirecting the output through an `os.Pipe` or temporary `os.File` so that we can stop reading it as soon as the command is canceled. 🤔

(In theory we could use the new `WaitDelay` field to bound the time spent waiting, but `WaitDelay` is not intended for cases where draining the buffer after a successful command might take longer than we're willing to wait upon cancellation.)

**Comment by findleyr** [comment]
> Thinking about this some more: if the process is already finished but Wait is taking longer than a second, it's probably spending that time ingesting a lot of (presumably buffered) output from the command's stdout and/or stderr.

That's an interesting idea. 10s is an absurdly long time for this to take though.

But your CL should be an improvement anyway.

**Comment by bcmills** [comment]
At present I believe that the `darwin` bug comprises this cluster of `watchflakes` issues:
- #60108
- #59995
- #59892
- #60099
- #54461

**Comment by findleyr** [comment]
Here's an interesting hang that wasn't on Darwin (linux-s390x-ibm):
https://build.golang.org/log/760ddd8386c57e95d80a543849ac2f344521dd07

**Comment by bcmills** [comment]
`linux-s390x-ibm` is a little suspect at the moment because of other timeout problems (#60413, #60109). I don't plan to look too much into that one.

**Comment by bcmills** [comment]
I have filed a new issue for the `darwin` hangs, since they seem to be much more general than either `gopls` or `cmd/go`: #63937.

**Comment by bcmills** [comment]
@pjweinb, I would suggest filing a separate issue for that — on a Google cloudtop in particular you may be running up against port exhaustion (see https://github.com/golang/go/issues/52545#issuecomment-1454157883).

**Comment by bcmills** [comment]
@golang/release, this issue has gotten very noisy because a bunch of these builders are just plain timing out. Could you please either prioritize a fix for those timeouts (maybe just set `GO_TEST_TIMEOUT_SCALE` really high for now?) or update the `watchflakes` pattern on this issue to exclude the misconfigured builders?

See: 
- #65311
- #64697
- #65040

**Comment by bcmills** [comment]
The recent spurious failures were #65845; marking as off-topic.

**Comment by findleyr** [comment]
This seems to be occurring more frequently now, perhaps due to telemetry? @matloob we should investigate.

**Comment by findleyr** [comment]
Here's a SIGQUIT dump from a hanging Go routine. FYI @matloob @samthanawalla this goroutine was hanging for a minute when invoked from one of our test repos (which should be very small).

https://logs.chromium.org/logs/golang/buildbucket/cr-buildbucket/8724574916972056209/+/u/step/21/log/2

**Comment by adonovan** [comment]
So, we have:
- modindex.dirHash -- stats every element of a package directory
- Package.Resolve -- calls ResolveImport for each import of a package
- runList -- calls p.Resolve(imports) for each package named by args, passing the list of its imports

All looks reasonable; in particular it doesn't appear to be cache trimming, nor has some kind of recursive file walk escaped from the intended tree. Some possibilities:
- the file system is oversubscribed from too much I/O; it would be nice to know the concurrency level;
- there's a single huge directory; it would be nice to know which one

@samthanawalla  It would be helpful to augment modindex.dirHash so that if it spends a long time in that loop (say, more than 30s), it starts logging pkgdir, the set of dirs, and the current concurrency level.


```
goroutine 1 gp=0xc000002380 m=nil [runnable]:
...
os.Lstat({0xc000800000, 0x3e})
	os/stat.go:26 +0x2c fp=0xc0000366a8 sp=0xc000036688 pc=0x50166c
os.(*unixDirent).Info(0xc1de0a8be455ef51?)
	os/file_unix.go:480 +0x57 fp=0xc0000366f0 sp=0xc0000366a8 pc=0x4ff5f7
cmd/go/internal/modindex.dirHash({0xc0001422d0, 0x2c}, {0xc000142300, 0x30})
	cmd/go/internal/modindex/read.go:111 +0x2af fp=0xc000036840 sp=0xc0000366f0 pc=0x81a94f
cmd/go/internal/modindex.openIndexPackage.func1()
	cmd/go/internal/modindex/read.go:216 +0x85 fp=0xc000036938 sp=0xc000036840 pc=0x81b325
cmd/internal/par.(*ErrCache[...]).Do.func1()
	cmd/internal/par/work.go:119 +0x13 fp=0xc000036948 sp=0xc000036938 pc=0x8258f3
cmd/internal/par.(*Cache[...]).Do(0xc737e0, {{0xc0001422d0, 0x2c}, {0xc000142300, 0x30}}, 0xc0000369f8)
	cmd/internal/par/work.go:160 +0x122 fp=0xc0000369c0 sp=0xc000036948 pc=0x825b02
cmd/internal/par.(*ErrCache[...]).Do(0xc000114460?, {{0xc0001422d0, 0x2c}, {0xc000142300, 0x30}}, 0x0?)
	cmd/internal/par/work.go:118 +0x46 fp=0xc000036a20 sp=0xc0000369c0 pc=0x8258a6
cmd/go/internal/modindex.openIndexPackage({0xc0001422d0?, 0xc000114460?}, {0xc000142300?, 0xc000114460?})
	cmd/go/internal/modindex/read.go:214 +0x98 fp=0xc000036ab0 sp=0xc000036a20 pc=0x81b258
cmd/go/internal/modindex.GetPackage({0xc0001422d0, 0x2c}, {0xc000142300, 0x30})
	cmd/go/internal/modindex/read.go:146 +0x14d fp=0xc000036af8 sp=0xc000036ab0 pc=0x81ad4d
cmd/go/internal/modindex.IsStandardPackage({0xc0001480c0, 0x28}, {0xb49338, 0x2}, {0x2d7091c1d, 0x3})
	cmd/go/internal/modindex/read.go:687 +0x20a fp=0xc000036bc8 sp=0xc000036af8 pc=0x81fbca
cmd/go/internal/modload.findStandardImportPath({0x2d7091c1d, 0x3})
	cmd/go/internal/modload/build.go:43 +0x67 fp=0xc000036c38 sp=0xc000036bc8 pc=0x82d367
cmd/go/internal/modload.Lookup({0x2d5126e0e?, 0xc000036cb8?}, 0x85?, {0x2d7091c1d?, 0xa6afe0?})
	cmd/go/internal/modload/load.go:868 +0x88 fp=0xc000036c80 sp=0xc000036c38 pc=0x859d48
cmd/go/internal/load.resolveImportPath({0x2d7091c1d, 0x3}, {0x2d5126e0e?, 0x47?}, {0xc000334100?, 0x9?}, {0xc0001480c0?, 0x0?}, 0x8?)
	cmd/go/internal/load/pkg.go:1174 +0x4e fp=0xc000036ce0 sp=0xc000036c80 pc=0x8a5a8e
cmd/go/internal/load.ResolveImportPath(0xc00017d600?, {0x2d7091c1d?, 0x2d708d1b2?})
	cmd/go/internal/load/pkg.go:1169 +0x69 fp=0xc000036d38 sp=0xc000036ce0 pc=0x8a5a09
cmd/go/internal/load.(*Package).Resolve(0xc00040a008, {0xc000005408, 0x39, 0xc000405080?})
	cmd/go/internal/load/pkg.go:365 +0x12a fp=0xc000036eb8 sp=0xc000036d38 pc=0x8a036a
cmd/go/internal/list.runList({0xc6e940, 0x10a1100}, 0xc0001a2120?, {0xc00018c0b0, 0x2, 0x2})
	cmd/go/internal/list/list.go:747 +0x187b fp=0xc0000379c8 sp=0xc000036eb8 pc=0x96f3bb
...
```

**Comment by findleyr** [comment]
FWIW, I think we should wait for a hang on Darwin or even NetBSD (or more on OpenBSD) before we draw too many conclusions. The OpenBSD builder tends to be incredibly slow, and so this may be a red herring.

**Comment by adonovan** [comment]
The last failure was on Darwin, and shows a call to os.RemoveAll (on behalf of `cmd/go/internal/work.(*Builder).Close`) taking a long time. It's unclear whether it is making progress but very slowly, or completely stuck.

```
os.removeAll({0xc001ec1000, 0x35})
	os/removeall_at.go:48 +0x227 fp=0xc0006c9098 sp=0xc0006c8fe0 pc=0xcad1e27
os.RemoveAll({0xc001ec1000?, 0x0?})
	os/path.go:73 +0x18 fp=0xc0006c90b8 sp=0xc0006c9098 pc=0xcad16d8
cmd/go/internal/work.(*Builder).Close.RemoveAll.removeAll.func1()
	cmd/go/internal/robustio/robustio_flaky.go:88 +0x1b fp=0xc0006c90e8 sp=0xc0006c90b8 pc=0xced709b
cmd/go/internal/robustio.retry(0xc0006c91d8)
	cmd/go/internal/robustio/robustio_flaky.go:29 +0xad fp=0xc0006c9168 sp=0xc0006c90e8 pc=0xcd96d2d
cmd/go/internal/robustio.removeAll(...)
	cmd/go/internal/robustio/robustio_flaky.go:87
cmd/go/internal/robustio.RemoveAll(...)
	cmd/go/internal/robustio/robustio.go:38
cmd/go/internal/work.(*Builder).Close(0xc000ae7290)
	cmd/go/internal/work/action.go:332 +0x1a5 fp=0xc0006c9228 sp=0xc0006c9168 pc=0xced6fa5
```

**Comment by Oguidan** [comment]
> At present I believe that the `darwin` bug comprises this cluster of `watchflakes` issues: [edit: I have pulled this list out to its own issue as [#63937](https://github.com/golang/go/issues/63937).)

But this is only on some processor architecture
